uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/AbsenceType.php

```diff
@@ -0,0 +1,206 @@
+<?php
+
+class CRM_HRLeaveAndAbsences_BAO_AbsenceType extends CRM_HRLeaveAndAbsences_DAO_AbsenceType {
+
+  const EXPIRATION_UNIT_DAYS = 1;
+  const EXPIRATION_UNIT_MONTHS = 2;
+  const EXPIRATION_UNIT_YEARS = 3;
+
+  const REQUEST_CANCELATION_NO = 1;
+  const REQUEST_CANCELATION_ALWAYS = 2;
+  const REQUEST_CANCELATION_IN_ADVANCE_OF_START_DATE = 3;
+
+  /**
+   * Create a new AbsenceType based on array-data
+   *
+   * @param array $params key-value pairs
+   * @return CRM_HRLeaveAndAbsences_DAO_AbsenceType|NULL
+   **/
+  public static function create($params) {
+    $className = 'CRM_HRLeaveAndAbsences_DAO_AbsenceType';
+    $entityName = 'AbsenceType';
+    $hook = empty($params['id']) ? 'create' : 'edit';
+
+    CRM_Utils_Hook::pre($hook, $entityName, CRM_Utils_Array::value('id', $params), $params);
+    self::validateParams($params);
+    if(isset($params['is_default']) && $params['is_default']) {
+      self::unsetDefaultTypes();
+    }
+    $instance = new $className();
+    $instance->copyValues($params);
+    $instance->save();
+    CRM_Utils_Hook::post($hook, $entityName, $instance->id, $instance);
+
+    return $instance;
+  }
+
+  public static function getRequestCancelationOptions() {
+     return [
+         self::REQUEST_CANCELATION_NO                       => ts('No'),
+         self::REQUEST_CANCELATION_ALWAYS                   => ts('Yes - always'),
+         self::REQUEST_CANCELATION_IN_ADVANCE_OF_START_DATE => ts('Yes - in advance of the start date of the leave')
+     ];
+  }
+
+  public static function getExpirationUnitOptions() {
+    return [
+        self::EXPIRATION_UNIT_DAYS   => ts('Days'),
+        self::EXPIRATION_UNIT_MONTHS => ts('Months'),
+        self::EXPIRATION_UNIT_YEARS  => ts('Years')
+    ];
+  }
+
+  private static function unsetDefaultTypes() {
+    $tableName = self::getTableName();
+    $query = "UPDATE $tableName SET is_default = 0 WHERE is_default = 1";
+    CRM_Core_DAO::executeQuery($query);
+  }
+
+  private static function validateParams($params) {
+    if(!empty($params['add_public_holiday_to_entitlement'])) {
+      self::validateAddPublicHolidayToEntitlement();
+    }
+
+    if (!empty('allow_request_cancelation') &&
+        !array_key_exists($params['allow_request_cancelation'], self::getRequestCancelationOptions())
```

Comment:
if (!empty('allow_request_cancelation') should be if (!empty($params['allow_request_cancelation'])